src/database.ts

```diff
@@ -0,0 +1,100 @@
+import { type Result } from "./registry/index.js";
+import {
+  type Config,
+  type AutoWaitConfig,
+  extractBaseUrl,
+  type ChainName,
+  getBaseUrl,
+  type Signal,
+  type Signer,
+  normalize,
+} from "./helpers/index.js";
+import { Statement } from "./statement.js";
+
+export class Database<D = unknown> {
+  readonly config: Config & Partial<AutoWaitConfig>;
+
+  constructor(config: Config & Partial<AutoWaitConfig> = {}) {
+    this.config = config;
+  }
+
+  static async readOnly(chainNameOrId: ChainName | number): Promise<Database> {
+    const baseUrl = getBaseUrl(chainNameOrId);
+    return new Database({ baseUrl });
+  }
+
+  static async forSigner(signer: Signer): Promise<Database> {
+    const baseUrl = await extractBaseUrl({ signer });
+    return new Database({ signer, baseUrl });
+  }
+
+  prepare<T = D>(sql: string): Statement<T> {
```

Comment:
The primary method for this Database interface. I'll discuss more in the Statement class what this means, but TL;DR is that we are currently "faking" prepared statements on the client-side to a) mimic the D1 API, and b) prepare for prepared statements (haha) when they actually land in Tableland.